Add some admin facing docs on data tables (#17585)
galaxyproject/galaxy: doc/source/admin/data_tables.md

# Tool data

Galaxy stores tool data in the path defined by `tool_data_path` (by default `tool-data/`).

> It's possible to to separate tool data of shed installed tools by setting (`shed_tool_data_path`).

Reviewer: ```suggestion
It's possible to separate tool data of ToolShed-installed tools by setting `shed_tool_data_path`.
```

Tool data consists of:

1. the actual data

> 2. a tool data table
> 3. tool data config files

Reviewer: Since you can have tool data with only a loc file and no tool data table config file, I think the following is more in line with the Galaxy lingo (see e.g. https://galaxyproject.org/admin/tools/data-tables/ ).

```suggestion
2. loc files
3. tool data table config files
```

Author: So the term *data table* should be defined as the set of unique data defined by the loc files that are referred to by config entries using the same name?

Maybe we also want to introduce the historic perspective for tool bound loc filled? For a better understanding...

Reviewer: Basically the data table is the `tool_data_table_conf.xml` file defining how to assign names to the columns in the `loc` file, which allowed Galaxy tool developers to move from using https://docs.galaxyproject.org/en/latest/dev/schema.html#from-file to https://docs.galaxyproject.org/en/latest/dev/schema.html#from-data-table .

Author: OK. Should we then also use the term *data table* for `tool_data_table_conf.xml` (instead of *tool data table config*)?

Reviewer: I think tool data table config is fine.

## Tool data

This is the actual data that is stored by default in `tool_data_path`. It may be favorable to store the
actual tool data in a separate folder. For manually managed tool data this can be achieved by simply
storing the data in another folder. For data that is added by data managers this can be achieved by
setting `galaxy_data_manager_data_path`.

> ## Tool data tables

Reviewer: ```suggestion
## Loc files
```

> In order to make tool data usable from Galaxy tools so called tool data tables are used. 

Reviewer: ```suggestion
In order to make tool data usable from Galaxy tools so called loc files are used. 
```

Those are tabular (by default tab separated) files with the extension `.loc`.

> Besides the actual paths the entries can contain, e.g. IDs, names, or other

Reviewer: ```suggestion
Besides the actual paths, the entries can contain IDs, names, or other information
```

that can be used in tools to select reference data. The paths should be given as absolute paths,
but can also be given relative to the Galaxy root dir.

> By default tool data tables are installed in `tool_data_path` (where also built-in tool data tables

Reviewer: ```suggestion
By default loc files are installed in `tool_data_path` (where also built-in loc files
```

are stored). By setting `shed_tool_data_path` this can be separated.

> ## Tool data table config

Reviewer: ```suggestion
## Tool data table config files
```

    <columns>value, dbkey, name, path</columns>
    <file path="bwa_index_color.loc" />
</table>
```

- `table`: `name`, `comment_char` (default `"#"`), `separator` (default `"\t"`), `allow_duplicate_entries` (default `True`), `empty_field_value` (default `""`)
- `columns`: a comma separated list of column names
- `file`: `path` (alternatively `url`, `from_config`)

Tool data table definitions for tools installed from a toolshed have an additional
element `tool_shed_repository` and sub-tags `tool_shed`
`repository_name`, `repository_owner`, `installed_changeset_revision`, e.g.:

```xml
<table name="plasmidfinder_database" comment_char="#">
    <columns>value, name, date, path</columns>
    <file path="/home/berntm/projects/galaxy/tool-data/toolshed.g2.bx.psu.edu/repos/iuc/plasmidfinder/7075b7a5441b/plasmidfinder_database.loc.sample"/>
    <tool_shed_repository>
        <tool_shed>toolshed.g2.bx.psu.edu</tool_shed>
        <repository_name>plasmidfinder</repository_name>
        <repository_owner>iuc</repository_owner>
        <installed_changeset_revision>7075b7a5441b</installed_changeset_revision>
    </tool_shed_repository>
</table>
```

> The file path points to a data table (i.e. a `.loc` file) and can be given

Reviewer: ```suggestion
The file path points to a loc file and can be given
```

relative (to the `tool_data_path`) or absolute. If a given relative path does
not exist also the base name is checked (many tools use something like
`tool-data/xyz.loc` and store example `loc` files in a `tool-data/` directory in
the tool repository).
Currently also `.loc.sample` may be used in case the specified `.loc` is absent.

Tool data table config files: 

- `tool_data_table_config_path`: by default `tool_data_table_conf.xml` in Galaxy's `config/` directory.
- `shed_tool_data_table_config`: by default `shed_tool_data_table_conf.xml` in
Galaxy's `config/` directory. This file lists all tool data tables of tools
installed from a toolshed. Note that the entries are versioned, i.e. there is a
separate entry for each tool and tool version. These content of the tool data
tables are merged when they are loaded.

When a new tool is installed that uses a data table a new entry is added to
`shed_tool_data_table_config` and a `.loc` file is placed in a versioned
subdirectory in `tool_data_path` (in a subdirectory that has the name of the

> toolshed). By default thus is `tool-data/toolshed.g2.bx.psu.edu/`. Note that

Reviewer: ```suggestion
toolshed). By default this is `tool-data/toolshed.g2.bx.psu.edu/`. Note that
```

I guess?

> these directories will also contain tool data config files, but they are unused.

Reviewer: ```suggestion
these directories will also contain tool data table config files, but they are unused.
```
